test/src/integration-query-condition.cc

```diff
@@ -0,0 +1,1386 @@
+
+  for (int r = 7; r <= 14; ++r) {
+    for (int c = 7; c <= 14; ++c) {
+      int og_i = ((r - 1) * 20) + (c - 1);
+      if (og_i % 2 == 0) {
+        int i = ((r - 7) * 4) + ((c - 7) / 2);
+        REQUIRE(a_data_read_2[i] == 1);
+        REQUIRE(a_data_read_2[i] == a_data_read[og_i]);
+        REQUIRE(
+            fabs(b_data_read_2[i] - b_data_read[og_i]) <
+            std::numeric_limits<float>::epsilon());
+      }
+    }
+  }
+
+  query2.finalize();
+  array2.close();
+
+  if (vfs.is_dir(array_name))
+    vfs.remove_dir(array_name);
+}
+
+TEST_CASE(
+    "Test read for sparse arrays with ranges across tiles on both dimensions, "
+    "with complex query condition",
+    "[query][query-condition][apply]") {
+  std::srand(static_cast<uint32_t>(time(0)));
+  Context ctx;
+  VFS vfs(ctx);
+
+  if (vfs.is_dir(array_name))
+    vfs.remove_dir(array_name);
+
+  std::vector<int> a_data_read(400);
+  std::vector<float> b_data_read(400);
+
+  setup(ctx, TILEDB_SPARSE, false, a_data_read, b_data_read);
+
+  Array array2(ctx, array_name, TILEDB_READ);
+  Query query2(ctx, array2);
+
+  QueryCondition qc1(ctx);
+  float val1 = 4.0f;
+  qc1.init("b", &val1, sizeof(float), TILEDB_LT);
+
+  QueryCondition qc2(ctx);
+  float val2 = 3.7f;
+  qc2.init("b", &val2, sizeof(float), TILEDB_LE);
+
+  QueryCondition qc3(ctx);
+  float val3 = 3.3f;
+  qc3.init("b", &val3, sizeof(float), TILEDB_GE);
+
+  QueryCondition qc4(ctx);
+  float val4 = 3.4f;
+  qc4.init("b", &val4, sizeof(float), TILEDB_NE);
+
+  QueryCondition qc5 = qc1.combine(qc2, TILEDB_AND);
+  QueryCondition qc6 = qc5.combine(qc3, TILEDB_AND);
+  QueryCondition qc = qc6.combine(qc4, TILEDB_AND);
+
+  std::vector<int> a_data_read_2(64);
+  std::vector<float> b_data_read_2(64);
+  int range[] = {7, 14};
+  query2.add_range("rows", range[0], range[1])
+      .add_range("cols", range[0], range[1]);
+  query2.set_layout(TILEDB_ROW_MAJOR)
+      .set_data_buffer("a", a_data_read_2)
+      .set_data_buffer("b", b_data_read_2);
+  query2.set_condition(qc);
+  query2.submit();
+  auto table2 = query2.result_buffer_elements();
+
+  REQUIRE(table2.size() == 2);
+  REQUIRE(table2["a"].first == 0);
+  REQUIRE(table2["a"].second == 8);
+  REQUIRE(table2["b"].first == 0);
+  REQUIRE(table2["b"].second == 8);
+
+  for (int r = 7; r <= 14; ++r) {
+    for (int c = 7; c <= 14; ++c) {
+      int og_i = ((r - 1) * 20) + (c - 1);
+      if (og_i % 8 == 4) {
+        int i = r - 7;
+        REQUIRE(a_data_read_2[i] == 1);
+        REQUIRE(a_data_read_2[i] == a_data_read[og_i]);
+        REQUIRE(
+            fabs(b_data_read_2[i] - b_data_read[og_i]) <
+            std::numeric_limits<float>::epsilon());
+      }
+    }
+  }
+
+  query2.finalize();
+  array2.close();
+
+  if (vfs.is_dir(array_name))
+    vfs.remove_dir(array_name);
+}
+
+TEST_CASE(
+    "Test read for sparse arrays (dups allowed) with no range, with query "
+    "condition",
+    "[query][query-condition][apply-sparse]") {
```

Comment:
Are all the [apply-sparse] tests exactly the same as the [apply] ones except a few parameters that we could change in a catch2 test variation?

Reply:
Yes, but I think Isaiah expressed concern of having difficulties debugging if we didn't separate the two concerns. I think his point is valid and think it's better for the test suite to have specifically different test cases for each apply variation for easier debugging, despite the additional cost of repetitive code.

Reply:
Isn't that what SECTION is for?

Reply:
I originally did the tests with sections but apparently this would make the tests harder to debug.

Reply:
@ihnorton Could you expand a little on why this makes in harder to debug? I would still prefer to use SECTION here.

Reply:
The original implementation had three TEST_CASE with SECTION blocks that looked fairly distinct to me, so separating into different cases with setup code factored out seemed like an improvement. If there is variation to capture with SECTION (or GENERATOR), and avoid repetition, then please do use those -- but with helper functions for common code when it's not a burden to do so (like `create_array` added in the recent commit 👍).

Reply:
These tests are quite doable with SECTION but I am unsure of what's being asked here--are we going back to 3 test cases with different sections? Or continuing onto 18 test cases?

Reply:
I think the suggestion is to try to reduce duplicated code by putting common setup/cleanup code into helper functions and then using SECTION to capture/specify differences.

Reply:
@abigalekim I think I'm ok with the tests as is... Especially that at some point apply is going away.

Reply:
Hello, I have gone back to the section layout and am using SECTION for each method. Especially since apply is going away soon I think this makes the most sense.

Reply:
@abigalekim Ok I see what happened here. I don't like this version with the sections, the previous one was better. 
Looking at the previous version, my comment was to try to combine for example these two test cases as they seemed the same exact code, but one for apply, and one for apply-sparse:
TEST_CASE(
    "Test read for sparse arrays with added range within a tile, with query "
    "condition",
    "[query][query-condition][apply]") {

and

TEST_CASE(
    "Test read for sparse arrays (dups allowed) with added range within a "
    "tile, with query condition",
    "[query][query-condition][apply-sparse]") {